input-profile.js

```diff
@@ -0,0 +1,194 @@
+        },
+        customProfileFolder: {
+            type: Type.String,
+        },
+    };
+
+    init() {
+        this.gamepadObjects = {};
+        this.modelLoaded = false;
+        this.tempVec = vec3.create();
+    }
+
+    start() {
+        if (this.engine.xrSession) {
+            this.engine.xrSession.addEventListener(
+                'inputsourceschange',
+                this.onInputSourcesChange.bind(this)
+            );
+        } else {
+            this.engine.onXRSessionStart.add((session) => {
+                session.addEventListener(
+                    'inputsourceschange',
+                    this.onInputSourcesChange.bind(this)
+                );
+            });
+        }
+
+        this.handedness = ['left', 'right'][this.handednessIndex];
+    }
+
+    update() {
+        if (this.modelLoaded === true) {
+            this.mapGamepadInput();
+        }
+    }
+
+    onInputSourcesChange(event) {
+        event.added.forEach((xrInputSource) => {
+            const profile = this.customProfileFolder || xrInputSource.profiles[0];
+            /** load Profile json */
+            /** Note: if you are providing a custom window.ProfileJSON provide its path in the
+             * customProfileFolder property of the input-profile component.
+             */
+            this.url = this.path + profile + '/profile.json';
+            console.log('loading Profile ' + this.url);
+            fetch(this.url)
+                .then((res) => res.json())
+                .then((out) => {
+                    window.ProfileJSON = out;
+                    const consoleText = this.modelLoaded
+                        ? 'Profile loaded'
+                        : 'Reloaded Profile for gamepad mapping ';
+                    console.log(consoleText);
+                })
+                .catch(console.error);
+
+            if (this.handedness == xrInputSource.handedness) {
+                this.gamepad = xrInputSource.gamepad;
+                /** load controllerModel **/
+                const assetPath = this.path + profile + '/' + this.handedness + '.glb';
+                if (!this.modelLoaded) {
+                    this.engine.scene
+                        .append(assetPath)
+                        .then((obj) => {
+                            obj.parent = this.object;
+                            obj.setTranslationLocal([0, 0, 0]);
+                            this.getGamepadObjectsFromProfile(window.ProfileJSON, obj);
+                            this.modelLoaded = true;
+                            console.log('model loaded to the scene');
+                        })
+                        .catch(console.error);
+                }
+            }
+        });
+    }
+
+    getGamepadObjectsFromProfile(profile, obj) {
+        //initialise components
+        const components = profile['layouts'][this.handedness]['components'];
```

Comment:
`profile.layouts`. Do not use strings when you know what to index.

Reply:
since iam reusing those tags (it can be changed as well , json from inputprofile github repo )
iam considering something like 

const layoutsKey = 'layouts';
const componentsKey = 'components';

const components = profile[layoutsKey][this.handedness][componentsKey];

Reply:
If they might change yes, this way is good. For constants on top-level, we use upper camel case: `ComponentsKey`

Reply:
although it give control over custom profile.json , I think for now let us stick with your suggestion `profile.layouts` since the json is less likely to change and instruct users to official guide to format their custom profiles.